File: src/main/java/com/polygloat/dtos/PublicConfigurationDTO.java
package com.polygloat.dtos;

import com.polygloat.configuration.AppConfiguration;
import lombok.Getter;

public class PublicConfigurationDTO {
    @Getter
    private boolean authentication;

    @Getter
    private AuthMethodsDTO authMethods;

    @Getter
    private boolean passwordResettable;

    @Getter
    private boolean allowRegistrations;

    public PublicConfigurationDTO(AppConfiguration configuration) {
        this.authentication = configuration.isAuthentication();
        if (authentication) {
            authMethods = new AuthMethodsDTO(new GithubPublicConfigDTO(configuration.getGithubClientId()));
        }
        passwordResettable = configuration.isNativeAuth();
        allowRegistrations = configuration.isAllowRegistrations();
    }

    public static class AuthMethodsDTO {
        @Getter
        private GithubPublicConfigDTO github;

        public AuthMethodsDTO(GithubPublicConfigDTO github) {
            this.github = github;
        }
    }

    public static class GithubPublicConfigDTO {
        @Getter
        private boolean enabled;

        @Getter
        private String clientId;

        public GithubPublicConfigDTO(String clientId) {
            this.clientId = clientId;
            this.enabled = clientId != null && !clientId.isEmpty();
        }
    }
}

File: src/main/java/io/tolgee/dtos/PublicConfigurationDTO.kt
package io.tolgee.dtos

import io.swagger.v3.oas.annotations.media.Schema
import io.tolgee.configuration.tolgee.TolgeeProperties

@Suppress("MemberVisibilityCanBePrivate", "unused")
class PublicConfigurationDTO(
        @Schema(hidden = true)
        properties: TolgeeProperties) {
    val isAuthentication: Boolean = properties.authentication.enabled
    var authMethods: AuthMethodsDTO? = null
    val isPasswordResettable: Boolean
    val isAllowRegistrations: Boolean
    val screenshotsUrl = properties.screenshotsUrl
    val maxUploadFileSize = properties.maxUploadFileSize
    val clientSentryDsn = if (properties.sentry.enabled) properties.sentry.clientDsn else null
    val needsEmailVerification = properties.authentication.needsEmailVerification

    class AuthMethodsDTO(val github: GithubPublicConfigDTO)
    class GithubPublicConfigDTO(clientId: String?) {
        val isEnabled: Boolean = clientId != null && clientId.isNotEmpty()
    }

    init {
        if (isAuthentication) {
            authMethods = AuthMethodsDTO(GithubPublicConfigDTO(properties.authentication.github.clientId))
        }
        isPasswordResettable = properties.authentication.nativeEnabled
        isAllowRegistrations = properties.authentication.registrationsAllowed
    }
}
